<a href="https://colab.research.google.com/github/RafaelLVaz/Checkpoint-02-Casos-de-Uso-de-ML---Energia-E-lica-e-Energia-Solar/blob/main/001_Parte2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Link do Kaggle: https://www.kaggle.com/datasets/pcbreviglieri/smart-grid-stability?resource=download

## Carregamento de dados

In [ ]:
import pandas as pd

df = pd.read_csv('/content/smart_grid_stability_augmented.csv')

display(df.head())

tau1      tau2      tau3      tau4        p1        p2        p3  \
0  2.959060  3.079885  8.381025  9.780754  3.763085 -0.782604 -1.257395   
1  9.304097  4.902524  3.047541  1.369357  5.067812 -1.940058 -1.872742   
2  8.971707  8.848428  3.046479  1.214518  3.405158 -1.207456 -1.277210   
3  0.716415  7.669600  4.486641  2.340563  3.963791 -1.027473 -1.938944   
4  3.134112  7.608772  4.943759  9.857573  3.525811 -1.125531 -1.845975   

         p4        g1        g2        g3        g4      stab     stabf  
0 -1.723086  0.650456  0.859578  0.887445  0.958034  0.055347  unstable  
1 -1.255012  0.413441  0.862414  0.562139  0.781760 -0.005957    stable  
2 -0.920492  0.163041  0.766689  0.839444  0.109853  0.003471  unstable  
3 -0.997374  0.446209  0.976744  0.929381  0.362718  0.028871  unstable  
4 -0.554305  0.797110  0.455450  0.656947  0.820923  0.049860  unstable

## Preparação dos dados


In [ ]:
X = df.drop(['stabf', 'stab'], axis=1)
y = df['stabf']

display(X.head())
display(y.head())
display(X.info())

tau1      tau2      tau3      tau4        p1        p2        p3  \
0  2.959060  3.079885  8.381025  9.780754  3.763085 -0.782604 -1.257395   
1  9.304097  4.902524  3.047541  1.369357  5.067812 -1.940058 -1.872742   
2  8.971707  8.848428  3.046479  1.214518  3.405158 -1.207456 -1.277210   
3  0.716415  7.669600  4.486641  2.340563  3.963791 -1.027473 -1.938944   
4  3.134112  7.608772  4.943759  9.857573  3.525811 -1.125531 -1.845975   

         p4        g1        g2        g3        g4  
0 -1.723086  0.650456  0.859578  0.887445  0.958034  
1 -1.255012  0.413441  0.862414  0.562139  0.781760  
2 -0.920492  0.163041  0.766689  0.839444  0.109853  
3 -0.997374  0.446209  0.976744  0.929381  0.362718  
4 -0.554305  0.797110  0.455450  0.656947  0.820923

0    unstable
1      stable
2    unstable
3    unstable
4    unstable
Name: stabf, dtype: object

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60000 entries, 0 to 59999
Data columns (total 12 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   tau1    60000 non-null  float64
 1   tau2    60000 non-null  float64
 2   tau3    60000 non-null  float64
 3   tau4    60000 non-null  float64
 4   p1      60000 non-null  float64
 5   p2      60000 non-null  float64
 6   p3      60000 non-null  float64
 7   p4      60000 non-null  float64
 8   g1      60000 non-null  float64
 9   g2      60000 non-null  float64
 10  g3      60000 non-null  float64
 11  g4      60000 non-null  float64
dtypes: float64(12)
memory usage: 5.5 MB


None

In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

display(X_scaled[:5])

array([[-0.83537431, -0.7913157 ,  1.1417039 ,  1.65210422,  0.01739718,
         1.07940409, -0.01707764, -1.09254449,  0.45746733,  1.22001344,
         1.32162771,  1.57902551],
       [ 1.47829663, -0.12670458, -0.80310987, -1.4150464 ,  1.75212436,
        -1.5936194 , -1.43815827, -0.01157471, -0.40679143,  1.23035459,
         0.13542396,  0.93625515],
       [ 1.35709296,  1.31213878, -0.80349711, -1.47150732, -0.45849151,
         0.09825193, -0.06283908,  0.76096398, -1.3198525 ,  0.88129894,
         1.14659597, -1.51380272],
       [-1.65313833,  0.88228802, -0.27835319, -1.06090381,  0.28425038,
         0.51390262, -1.5910461 ,  0.58341479, -0.2873044 ,  1.64725049,
         1.47454278, -0.59175095],
       [-0.77154307,  0.86010756, -0.11166849,  1.68011561, -0.2980753 ,
         0.2874485 , -1.37634301,  1.60663641,  0.99222554, -0.25360995,
         0.48113288,  1.07906223]])

## Divisão dos dados


In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

## Treinamento dos modelos

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression

dt_model = DecisionTreeClassifier()
knn_model = KNeighborsClassifier()
lr_model = LogisticRegression()

dt_model.fit(X_train, y_train)
knn_model.fit(X_train, y_train)
lr_model.fit(X_train, y_train)

LogisticRegression()

## Avaliação dos modelos

In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix, f1_score

dt_pred = dt_model.predict(X_test)
knn_pred = knn_model.predict(X_test)
lr_pred = lr_model.predict(X_test)

dt_accuracy = accuracy_score(y_test, dt_pred)
dt_confusion = confusion_matrix(y_test, dt_pred)
dt_f1 = f1_score(y_test, dt_pred, pos_label='unstable')

knn_accuracy = accuracy_score(y_test, knn_pred)
knn_confusion = confusion_matrix(y_test, knn_pred)
knn_f1 = f1_score(y_test, knn_pred, pos_label='unstable')

lr_accuracy = accuracy_score(y_test, lr_pred)
lr_confusion = confusion_matrix(y_test, lr_pred)
lr_f1 = f1_score(y_test, lr_pred, pos_label='unstable')

print("Desempenho do modelo de árvore de decisão:")
print(f"Acurácia: {dt_accuracy:.4f}")
print("Matriz de confusão:")
display(dt_confusion)
print(f"F1-score: {dt_f1:.4f}")
print("-" * 30)

print("Desempenho do modelo KNN:")
print(f"Acurácia: {knn_accuracy:.4f}")
print("Matriz de confusão:")
display(knn_confusion)
print(f"F1-score: {knn_f1:.4f}")
print("-" * 30)

print("Desempenho do modelo de regressão logística:")
print(f"Acurácia: {lr_accuracy:.4f}")
print("Matriz de confusão:")
display(lr_confusion)
print(f"F1-score: {lr_f1:.4f}")

Desempenho do modelo de árvore de decisão:
Acurácia: 0.8975
Matriz de confusão:


array([[3731,  591],
       [ 639, 7039]])

F1-score: 0.9196
------------------------------
Desempenho do modelo KNN:
Acurácia: 0.8872
Matriz de confusão:


array([[3421,  901],
       [ 453, 7225]])

F1-score: 0.9143
------------------------------
Desempenho do modelo de regressão logística:
Acurácia: 0.8102
Matriz de confusão:


array([[3004, 1318],
       [ 960, 6718]])

F1-score: 0.8550


## Comparação e discussão

### **Resultados dos Modelos**

**Árvore de Decisão (Decision Tree)**
* **Acurácia:** 0.8975
* **F1-score:** 0.9196
* **Matriz de Confusão:**
    * Verdadeiro Negativo: 3731 | Falso Positivo: 591
    * Falso Negativo: 639 | Verdadeiro Positivo: 7039

---

**K-Nearest Neighbors (KNN)**
* **Acurácia:** 0.8872
* **F1-score:** 0.9143
* **Matriz de Confusão:**
    * Verdadeiro Negativo: 3421 | Falso Positivo: 901
    * Falso Negativo: 453 | Verdadeiro Positivo: 7225

---

**Regressão Logística (Logistic Regression)**
* **Acurácia:** 0.8102
* **F1-score:** 0.8550
* **Matriz de Confusão:**
    * Verdadeiro Negativo: 3004 | Falso Positivo: 1318
    * Falso Negativo: 960 | Verdadeiro Positivo: 6718

---

### **Discussão**

Com base nas novas métricas de avaliação, o modelo de Árvore de Decisão continua sendo o mais confiável para a tarefa de detectar instabilidade na rede elétrica. Ele alcançou a maior acurácia (0.8975) e o **maior F1-score (0.9196) entre os três modelos testados.

O F1-score é uma métrica crucial neste contexto, pois oferece um balanço entre precisão e sensibilidade (recall), sendo fundamental para avaliar a capacidade do modelo em identificar corretamente os casos de instabilidade, minimizando erros como falsos positivos e falsos negativos.

O modelo KNN apresentou um desempenho competitivo e muito próximo, com um F1-score de 0.9143, mas a ligeira superioridade da Árvore de Decisão a torna a escolha preferencial. A Regressão Logística, por sua vez, demonstrou um desempenho consideravelmente inferior, com um F1-score de 0.8550, mostrando-se menos adequada para este problema de classificação.